contracts/utils/structs/Heap.sol

```diff
@@ -0,0 +1,228 @@
+     *
+     * Note: All inserting and removal from a heap should always be done using the same comparator. Mixing comparator
+     * during the lifecycle of a heap will result in undefined behavior.
+     */
+    function pop(
+        Uint256Heap storage self,
+        function(uint256, uint256) view returns (bool) comp
+    ) internal returns (uint256) {
+        uint32 size = length(self);
+
+        if (size == 0) Panic.panic(Panic.EMPTY_ARRAY_POP);
+
+        uint32 last = size - 1; // could be unchecked (check above)
+
+        // get root location (in the data array) and value
+        uint32 rootIdx = self.data[0].index;
+        uint256 rootValue = self.data[rootIdx].value;
+
+        // if root is not the last element of the data array (that will get pop-ed), reorder the data array.
+        if (rootIdx != last) {
+            // get details about the value stored in the last element of the array (that will get pop-ed)
+            uint32 lastDataIdx = self.data[last].lookup;
+            uint256 lastDataValue = self.data[last].value;
+            // copy these values to the location of the root (that is safe, and that we no longer use)
+            self.data[rootIdx].value = lastDataValue;
+            self.data[rootIdx].lookup = lastDataIdx;
+            // update the tree node that used to point to that last element (value now located where the root was)
+            self.data[lastDataIdx].index = rootIdx;
+        }
+
+        // get last leaf location (in the data array) and value
+        uint32 lastIdx = self.data[last].index;
+        uint256 lastValue = self.data[lastIdx].value;
+
+        // move the last leaf to the root, pop last leaf ...
+        self.data[0].index = lastIdx;
+        self.data[lastIdx].lookup = 0;
+        self.data.pop();
+
+        // ... and heapify
+        _heapifyDown(self, last, 0, lastValue, comp);
+
+        // return root value
+        return rootValue;
+    }
+
+    /**
+     * @dev Insert a new element in the heap using the default comparator.
+     *
+     * Note: All inserting and removal from a heap should always be done using the same comparator. Mixing comparator
+     * during the lifecycle of a heap will result in undefined behavior.
+     */
+    function insert(Uint256Heap storage self, uint256 value) internal {
+        insert(self, value, Comparators.lt);
+    }
+
+    /**
+     * @dev Insert a new element in the heap using the provided comparator.
+     *
+     * Note: All inserting and removal from a heap should always be done using the same comparator. Mixing comparator
+     * during the lifecycle of a heap will result in undefined behavior.
+     */
+    function insert(
+        Uint256Heap storage self,
+        uint256 value,
+        function(uint256, uint256) view returns (bool) comp
+    ) internal {
+        uint32 size = length(self);
+        self.data.push(Uint256HeapNode({index: size, lookup: size, value: value}));
+        _heapifyUp(self, size, value, comp);
```

Comment:
is length is type(uint32).max, the insert works, but then the size is increassed, and the length function reverts. Maybe we want to check here and revert ?

That check would hewever have a cost, and its won't realistically be triggered.